com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/ide/contentAssist/AnnotationAwareFollowElementComputer.java

```diff
@@ -0,0 +1,78 @@
+ * Copyright (c) 2016 Avaloq Evolution AG and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Avaloq Evolution AG - initial API and implementation
+ *******************************************************************************/
+
+package com.avaloq.tools.ddk.xtext.generator.ide.contentAssist;
+
+import java.util.Collection;
+
+import org.eclipse.xtext.AbstractElement;
+import org.eclipse.xtext.GrammarUtil;
+import org.eclipse.xtext.ParserRule;
+import org.eclipse.xtext.RuleCall;
+import org.eclipse.xtext.ide.editor.contentassist.IFollowElementAcceptor;
+import org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElement;
+import org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementCalculator;
+import org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer;
+
+import com.avaloq.tools.ddk.xtext.parser.ISemanticPredicates;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+
+public class AnnotationAwareFollowElementComputer extends FollowElementComputer {
+  @Inject
+  private Provider<FollowElementCalculator> feCalculatorProvider;
+  @Inject
+  private ISemanticPredicates keywordPredicates;
+
+  @Override
+  public void computeFollowElements(final Collection<FollowElement> followElements, final IFollowElementAcceptor followElementAcceptor) {
+    // copy of super class but initialisation of calculator moved to protected method.
+    FollowElementCalculator calculator = feCalculatorProvider.get();
+    initialiseCalculator(calculator, followElementAcceptor);
+    for (FollowElement element : followElements) {
+      computeFollowElements(calculator, element);
+    }
+  }
+
+  protected void initialiseCalculator(final FollowElementCalculator calculator, final IFollowElementAcceptor followElementAcceptor) {
+    // copied from super class but calls to keyword rules are also accepted.
+    calculator.setAcceptor(element -> {
+      ParserRule rule = GrammarUtil.containingParserRule(element);
+      if (rule == null || !GrammarUtil.isDatatypeRule(rule) || isKeywordRuleCall(element) || isEnumLikeDatatypeRule(rule)) {
+        followElementAcceptor.accept(element);
+      }
+    });
+  }
+
+  private boolean isKeywordRuleCall(final AbstractElement element) {
+    if (element instanceof RuleCall) {
+      RuleCall ruleCall = (RuleCall) element;
+      return keywordPredicates.isKeywordRule(ruleCall.getRule().getName());
+    }
+    return false;
+  }
+
+  /**
+   * Determines whether a ParserRule is a datatype rule containing keywords only.
+   *
+   * @param rule
+   *          the rule to be tested
+   * @return true if datatype rule contains keywords only
+   */
+  private static boolean isEnumLikeDatatypeRule(final ParserRule rule) {

```

Comment:
This method is identical to the one in AnnotationAwareFollowElementCalculator, could we refactor so that the code is reused?

Reply:
Done